### DISCUSSION TITLE
cedrus264 dont show video at other side

### DISCUSSION THREAD

**Discussion by alexlansovich** [baresip/baresip#2608] [General]
I have use latest baresip (baresip v3.0.0) and cant get working encoder cedrus264.
I have hardware NanoPi Duo 2 with connected camera on OV5640 connector.
Was compiled ffmpeg with cedrus264 encoder, than was compiled baresip.
If i test `ffmpeg -t 30 -f v4l2 -channel 0 -video_size 1280x720 -i /dev/video0 -pix_fmt nv12 -r 30 -b:v 64k test.mp4`
i get more then 39 FPS in video.
When i set encoder cedrus264 in baresip - i also get 30 fps, but thereis no video on other side.
If i set vcodec_h264enc        libx264 - i see video, but is low FPS( about 15)
```
baresip -v
baresip v3.0.0 Copyright (C) 2010 - 2022 Alfred E. Heggestad et al.
Local network addresses:
       eth0:  192.168.88.3
       eth0:  fd91:2e1a:f627:4783:a8fd:43ff:fe24:e315
       eth0:  fe80::a8fd:43ff:fe24:e315
uag: add local address 192.168.88.3
ua: adding SIP CA file: /etc/ssl/certs/ca-certificates.crt
ua: adding SIP CA path: 
uag: add local address fd91:2e1a:f627:4783:a8fd:43ff:fe24:e315
uag: add local address fe80::a8fd:43ff:fe24:e315
ui: stdio
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: auconv
aufilt: auresamp
ausrc: alsa
auplay: alsa
vidcodec: H264
vidcodec: H264
vidcodec: H265
avcodec: using H.264 encoder 'cedrus264' -- Cedrus H.264 Encoder
avcodec: using H.264 decoder 'h264' -- H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
avcodec: using H.265 encoder 'libx265' -- libx265 H.265 / HEVC
avcodec: using H.265 decoder 'hevc' -- HEVC (High Efficiency Video Coding)
vidsrc: v4l2
uuid: loaded UUID b64babf5-b923-22b1-40bd-78722e1f082c from file /root/.baresip/uuid
medianat: stun
medianat: turn
medianat: ice
module: loading app account.so
ua: ua_register sip:1003@vpbx.xx.xx:5060
Populated 1 account
module: loading app contact.so
Populated 3 contacts
module: loading app debug_cmd.so
module: loading app menu.so
module: loading app netroam.so
Populated 2 audio codecs
Populated 2 audio filters
Populated 3 video codecs
Populated 0 video filters
baresip is ready.
1003@vpbx.xx.xx: selected for 1003-0xb709ee28
ua: incoming OPTIONS message from sip:Unknown@ip (ip:5060)
call: alloc with params laddr=192.168.88.3, af=AF_INET, use_rtp=0
call: use_video=1
video: stopping video source ..
video: stopping video display ..
audio: destroyed (started=0)
1003@vpbx.xx.xx: (prio 0) {0/UDP/v4} 200 OK (FPBX-16.0.40(18.16.0)) [1 binding]
All 1 useragent registered successfully! (30 ms)
```
test call log

```
ua: incoming OPTIONS message from sip:Unknown@ip (ip:5060)
call: alloc with params laddr=192.168.88.3, af=AF_INET, use_rtp=0
call: use_video=1
video: stopping video source ..
video: stopping video display ..
audio: destroyed (started=0)
call: got SDP answer (423 bytes)
1003@vpbx.xx.xx: Call answered: sip:1005@vpbx.xx.xx
alsa: stopping playback thread (default)
call: update media
stream: update 'audio'
stream: audio: starting RTCP with remote ip:15029
stream: update 'video'
stream: video: starting RTCP with remote ip:16487
audio: update
audio: Set audio decoder: PCMU 8000Hz 1ch
audio: start
audio: create auplay buffer [20 - 160 ms] [320 - 2560 bytes]
alsa: reset: srate=8000, ch=1, num_frames=160, pcmfmt=S16_LE
alsa: playback started (default)
audio: player started with sample format S16LE
audio: Set audio encoder: PCMU 8000Hz 1ch
audio: start
alsa: reset: srate=8000, ch=1, num_frames=160, pcmfmt=S16_LE
alsa: recording started (default) format=S16LE
audio: source started with sample format S16LE
audio tx pipeline:        alsa ---> aubuf ---> auconv ---> auresamp ---> PCMU
audio rx pipeline:        alsa <--- aubuf <--- auconv <--- auresamp <--- PCMU
video: update
Set video encoder: H264 packetization-mode=0 (1000000 bit/s, 30.00 fps)
avcodec: h264 encoder activated
avcodec: video encoder H264: 30.00 fps, 1000000 bit/s, pktsize=1280
video: start source
auresamp: resample encoder 8000/1 --> 8000/1
v4l2: /dev/video0: found valid V4L2 device (640 x 480) pixfmt=YU12
v4l2: Current input: ""
v4l2: current framerate is 30.00 fps
video: stopping video display ..
call: stream start (active=1)
audio: start
video: update
video: stopping video display ..
stream: audio: enable RTP from remote
stream: video: enable RTP from remote
1003@vpbx.xx.xx: Call established: sip:1005@vpbx.xx.xx
call: modify
call: update media
stream: update 'audio'
stream: audio: starting RTCP with remote ip:15029
stream: update 'video'
stream: video: starting RTCP with remote ip:16487
audio: update
video: update
video: stopping video display ..
call: got SDP answer (423 bytes)
1003@vpbx.xx.xx: Call answered: sip:1005@vpbx.xx.xx
call: update media
stream: update 'audio'
stream: audio: starting RTCP with remote ip:15029
stream: update 'video'
stream: video: starting RTCP with remote ip:16487
audio: update
video: update
video: stopping video display ..
stream: incoming rtp for 'audio' established, receiving from ip:15028
[VDPAU SUNXI] VE version 0x1680 opened.
auresamp: resample decoder 8000/1 --> 8000/1
[0:00:09] audio=64000/63978 video=43665/0 (bit/s) efps=29.6/0.0    
--- Video stream ---
 source started: yes
 display started: no
 tx: encode: H264 nv12
     source: v4l2 640 x 480, fps=30.00 frames=271
     skipc=0 sendq=0
     time = 8.994 sec
 rx: decode: none ???
     vidisp: none 0 x 0 frames=0
     n_keyframes=0, n_picup=0
     time = (not started)
--- Stream debug ---
 video dir=sendonly pt_enc=96
 local: 0.0.0.0:13040, remote: ip:16486/ip:16487
 mnat: (none) (connected=no)
 menc: (none) (secure=no)
 rx.enabled: yes
 tx.enabled: yes
RTP debug:
 Encode: seq=32364 ssrc=0x294392e2
----- RTCP Session: -----
  cname=sip:1003@vpbx.xx.xx:5060 SSRC=0x294392e2/692294370 rx=90000Hz
  member 0x7427fd79: lost=0 Jitter=32.9ms RTT=11.5ms
                 IP=ip:16486 psent=27 rcvd=31
  TX: packets=544, octets=48234
--- jitter buffer debug---
 running=0 min=5 cur=0 max=10 [frames]
 seq_put=0
sip:1005@vpbx.xx.xx: session closed: Connection reset by peer [104]
alsa: stopping recording thread (default)
alsa: stopping playback thread (default)
video: stopping video source ..
v4l2: stopping video source..
video: stopping video display ..
sip:1003@vpbx.xx.xx:5060: Call with sip:1005@vpbx.xx.xx terminated (duration: 13 secs)
video: stopping video source ..
video: stopping video display ..
audio: destroyed (started=0)
```

```
ffmpeg -hide_banner -encoders | grep 264
 V..... cedrus264            Cedrus H.264 Encoder (codec h264)
 V....D libx264              libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 (codec h264)
 V....D libx264rgb           libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 RGB (codec h264)
 V..... h264_v4l2m2m         V4L2 mem2mem H.264 encoder wrapper (codec h264)
```

```
ffmpeg version N-103905-gafd516bd9c Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 7 (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04)
  configuration: --prefix=/usr --enable-nonfree --enable-gpl --enable-libx264 --enable-libmp3lame --enable-libpulse --enable-libv4l2 --enable-libx265 --enable-shared --arch=armv7l --target-os=linux --enable-version3 --enable-vdpau
```

**Comment by alfredh** [comment]
can you please share relevant info such as config files:

https://github.com/baresip/baresip/wiki/Guidelines-for-reporting-Issues

**Comment by alexlansovich** [comment]
Is this info enough?
FFmpeg was compiled from https://github.com/twdragon/FFmpeg

Baresip latest:
baresip v3.1.0 Copyright (C) 2010 - 2022 Alfred E. Heggestad et al.
Local network addresses:
       eth0:  192.168.88.3
       eth0:  fd91:2e1a:f627:4783:a8fd:43ff:fe24:e315
       eth0:  fe80::a8fd:43ff:fe24:e315
ua: adding SIP CA file: /etc/ssl/certs/ca-certificates.crt
ua: adding SIP CA path:
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: auconv
aufilt: auresamp
ausrc: alsa
auplay: alsa
vidcodec: H264
vidcodec: H264
vidcodec: H265
avcodec: using H.264 encoder 'cedrus264' -- Cedrus H.264 Encoder
avcodec: using H.264 decoder 'h264' -- H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
avcodec: using H.265 encoder 'libx265' -- libx265 H.265 / HEVC
avcodec: using H.265 decoder 'hevc' -- HEVC (High Efficiency Video Coding)
vidsrc: v4l2
medianat: stun
medianat: turn
medianat: ice
Populated 1 account
Populated 3 contacts
Populated 2 audio codecs
Populated 2 audio filters
Populated 3 video codecs
Populated 0 video filters
baresip is ready.

root@NanoPi-Duo2:/root/.baresip# cat accounts | egrep -v "^\s*(#|$)"
<sip:1008@pbx.xxx.net:5060>;auth_pass=xxx;answermode=auto;sip_autoanswer=yes

root@NanoPi-Duo2:/root/.baresip# cat config | egrep -v "^\s*(#|$)"
poll_method		epoll		# poll, select, epoll ..
sip_cafile		/etc/ssl/certs/ca-certificates.crt
sip_tos			160
call_local_timeout	120
call_max_calls		4
call_hold_other_calls	yes
audio_player		alsa,default
audio_source		alsa,default
audio_alert		alsa,default
audio_level		no
ausrc_format		s16		# s16, float, ..
auplay_format		s16		# s16, float, ..
auenc_format		s16		# s16, float, ..
audec_format		s16		# s16, float, ..
audio_buffer		20-160		# ms
audio_buffer_mode	fixed		# fixed, adaptive
audio_silence		-35.0		# in [dB]
audio_telev_pt		101		# payload type for telephone-event
video_source		v4l2,/dev/video0
video_size		640x480
video_bitrate		1000000
video_fps		30.00
video_fullscreen	no
videnc_format           nv12
rtp_tos			184
rtp_video_tos		136
jitter_buffer_type	fixed		# off, fixed, adaptive
jitter_buffer_delay	5-10		# frames
rtp_stats		no
module_path		/usr/local/lib/baresip/modules
module			stdio.so
module			g711.so
module			auconv.so
module			auresamp.so
module			alsa.so
module			avcodec.so
module			v4l2.so
module			uuid.so
module			stun.so
module			turn.so
module			ice.so
module_app		account.so
module_app		contact.so
module_app		debug_cmd.so
module_app		menu.so
module_app		netroam.so
cons_listen		0.0.0.0:5555 # cons - Console UI UDP/TCP sockets
http_listen		0.0.0.0:8000 # httpd - HTTP Server
ctrl_tcp_listen		0.0.0.0:4444 # ctrl_tcp - TCP interface JSON
evdev_device		/dev/input/event0
opus_bitrate		28000 # 6000-510000
vumeter_stderr		yes
video_selfview		window # {window,pip}
avcodec_h264enc		cedrus264

**Comment by alfredh** [comment]
Can you please set up a working environment first between A and B ?
Start with libx264 even if it is slow.


After the basic working setup is working, start changing things one by one,
and see what breaks.


Also enable this config option:

```
rtp_stats               yes
```

And run baresip with `-v` verbose flag.


There is also a `vidloop` module that you can test:
https://github.com/baresip/baresip-apps/tree/main/modules/vidloop


Also some more advice:

1. Try to exclude the SIP server and do a direct Peer-to-peer call
2. Run Wireshark to check if packets are flowing
3. Check the remote endpoint, does it receive packets ?

**Comment by alexlansovich** [comment]
Sorry for the big font, I dont understand how to clear this.
Thank you for the information - how setup vidloop. Please add this to the baresip wiki.
==
I have test working environment.
using avcodec_h264enc libx264 and avcodec_h264dec h264 video is working, but 15 fps.
==
With vidloop command = baresip -e/vidloop
avcodec_h264enc libx264 and avcodec_h264enc cedrus264 showing video ad 30 FPS.
with command  = baresip -e"/vidloop h264" showing video 15 FPS.
Trying running vidloop with cedrus i get error, because baresip cant find encoder and decoder "cedrus".
I have test baresip -e"/vidloop cedrus", baresip -e"/vidloop cedrus264" - no success.
Also - i think vidloop should have option for the encoder and decoder. Because i have cedrus264 encoder, not decoder.
Result of test vidloop without command codec, but config using this i get 30 FPS
avcodec_h264enc cedrus264
avcodec_h264dec h264
baresip -e"vidloop" > def-cedrus.log
=
https://gist.github.com/alexlansovich/0df337f6d14cca251296f64265fd14f6
==
Result of test vidloop without command codec, but config using this i get 30 FPS
avcodec_h264enc libx264
avcodec_h264dec h264
baresip -e"vidloop" > def-x264.log
https://gist.github.com/alexlansovich/887e7169edd3eb02d06b83c1228b84c7
==
I have enable rtp_stats               yes, and run baresip -v. I Testing with only video for outgoing call.
with avcodec_h264enc libx264 - result is https://gist.github.com/alexlansovich/263c5fa72d61fed2bcc1c8c49e00bfad
I have video.
with avcodec_h264enc cedrus264 - result is https://gist.github.com/alexlansovich/a473f8b9ba4eabac64fe44ae81df7e28
I dont have video.
==
1. Try to exclude the SIP server and do a direct Peer-to-peer call - what i need to set in account config, and haw to make call?
2. Run Wireshark to check if packets are flowing = shouild i do this on the baresip side?
3. Check the remote endpoint, does it receive packets = how i can check this?
==
I make several tests. Baresip call to MAC linphone, Baresip call to Iphone several video APPS. Baresip call to other baresip.
Thank you for the support!

**Comment by alfredh** [comment]
it looks like the video is sent from your endpoint:


```
video           Transmit:     Receive:
packets:          28278            0
avg. bitrate:    4226.6          0.0  (kbit/s)
errors:               0            0
pkt.report:       26864         2701
lost:                 0            0
jitter:             3.6          0.4  (ms)
```


now we have to figure out if the packets are received by B.


Please use Wireshark and sniff on the network connection between A and B.


Also try direct peer-to-peer call w/o sip server, example SIP account:


```
<sip:user@10.0.0.1>;regint=0
```

**Comment by alexlansovich** [comment]
I have configure accouns on two devices like:
`device 1 - ip 192.168.88.2.
account config = <sip:user2@192.168.88.2>;regint=0`
`device 2 - ip 192.168.88.3.
account config = <sip:user3@192.168.88.3>;regint=0`

I make call from second device to first.
`d sip:user2@192.168.88.2` or `user2@192.168.88.2`
and this nothing.

Populated 1 account
module: loading app contact.so
Populated 3 contacts
module: loading app debug_cmd.so
module: loading app menu.so
module: loading app netroam.so
Populated 2 audio codecs
Populated 2 audio filters
Populated 3 video codecs
Populated 0 video filters
baresip is ready.
           sip:user2@192.168.88.2
user3@192.168.88.3: selected for request
call uri: sip:user2@192.168.88.2
call: alloc with params laddr=192.168.88.3, af=AF_INET, use_rtp=1
call: use_video=1
call: connecting to sip:user2@192.168.88.2..
call id: fa3c4e6d87e8c6a1
sip:user2@192.168.88.2: session closed: Connection timed out [110]
video: stopping video source ..
video: stopping video display ..
video: stopping video source ..
video: stopping video display ..
audio: destroyed (started=0)

I have make autoanswer but, thereis no call at other side.
`<sip:user2@192.168.88.2>;regint=0;answermode=auto;sip_autoanswer=yes`.

==

i have make test capture of traffic - how i can give you dump?

**Comment by alexlansovich** [comment]
Update!
I have found working mode. At freepbx (asterisk) at "General Settings" i switch off parametr "Strict RTP".
All sip extensions have option "nat" - "yes, force rport, comedia".
Video is working, but only two calls. After two calls baresip is core damped.
https://gist.github.com/alexlansovich/baf87f5f480acd3a9bc0178359ad2ff0

When parametr "Strict RTP" is on i see log in asterisk:
If i turn off this parametr - Not Acceptable Here -is not present in log and video started very fast.
-
-- Called SIP/1007
 -- SIP/1007-0000037a is ringing
    > 0x7feeb8038730 -- Strict RTP learning after remote address set to: 192.168.88.99:55947
    > 0x7feeb806d4c0 -- Strict RTP learning after remote address set to: 192.168.88.99:55949
 -- SIP/1007-0000037a answered SIP/1009-00000379
 -- Channel SIP/1007-0000037a joined 'simple_bridge' basic-bridge <cc578024-f119-4f87-a129-3fb9925b0ffa>
 -- Channel SIP/1009-00000379 joined 'simple_bridge' basic-bridge <cc578024-f119-4f87-a129-3fb9925b0ffa>
    > 0x7feeb8038730 -- Strict RTP qualifying stream type: audio
    > 0x7feeb8038730 -- Strict RTP switching source address to 195.x.x.32:55947
    > 0x7feeb00c8180 -- Strict RTP qualifying stream type: audio
    > 0x7feeb00c8180 -- Strict RTP switching source address to 195.x.x.32:19050
    > 0x7feeb806d4c0 -- Strict RTP qualifying stream type: video
    > 0x7feeb806d4c0 -- Strict RTP switching source address to 195.x.x.32:55949
    > 0x7feeb00d8830 -- Strict RTP qualifying stream type: video
    > 0x7feeb00d8830 -- Strict RTP switching source address to 195.x.x.32:17366
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
    > 0x7feeb00c8180 -- Strict RTP learning complete - Locking on source address 195.x.x.32:19050
    > 0x7feeb00d8830 -- Strict RTP learning complete - Locking on source address 195.x.x.32:17366
    > 0x7feeb8038730 -- Strict RTP learning complete - Locking on source address 195.x.x.32:55947
    > 0x7feeb806d4c0 -- Strict RTP learning complete - Locking on source address 195.x.x.32:55949
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
 -- Got SIP INFO response 488 "Not Acceptable Here" back from host '195.x.x.32:39521'
--
My network scheme is.
Sip helper on mikrotik is disabled.
![Network](https://github.com/baresip/baresip/assets/7941877/1c3588b1-c577-42c2-a697-5f426d86b3a4)

**Comment by alfredh** [comment]
this is not a bug in baresip. Please contact FreePBX and ask them about the error logs.

Please also try peer-to-peer calls to exclude FreePBX:

config:

```
sip_listen             0.0.0.0:5060    
```


moving to discussion.

**Comment by alexlansovich** [comment]
Thank you. Direct call is working.
But only two calls. after second call baresip is dying - Segmentation fault (core dumped) - https://gist.github.com/alexlansovich/baf87f5f480acd3a9bc0178359ad2ff0

How we can investigate this?
#pidof baresip = 8364

run 
#strace -f -p 8364

[pid  8364] epoll_ctl(3, EPOLL_CTL_DEL, 14, 0xbee2f130) = 0
[pid  8364] close(14)                   = 0
[pid  8364] epoll_ctl(3, EPOLL_CTL_DEL, 15, 0xbee2f138) = 0
[pid  8364] close(15)                   = 0
[pid  8364] munmap(0xb118e000, 460800)  = 0
[pid  8364] --- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0xf1} ---
[pid  8368] <... futex resumed>)        = ?
[pid  8367] <... futex resumed>)        = ?
[pid  8366] <... futex resumed>)        = ?
[pid  8365] <... futex resumed>)        = ?
[pid  8368] +++ killed by SIGSEGV +++
[pid  8367] +++ killed by SIGSEGV +++
[pid  8366] +++ killed by SIGSEGV +++
[pid  8365] +++ killed by SIGSEGV +++
+++ killed by SIGSEGV +++

**Comment by alexlansovich** [comment]
I have run gdb to find problem and see that this is encoder problem. So we will try to correct cedrus encoder in FFMPEG.
==
Thread 12 "v4l2" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb2f47460 (LWP 16008)]
memlist_add (mem=0xb1c71780) at libavcodec/arm/sunxi/ve.c:97
97			m = m->next;
